launch_ros/launch_ros/utilities/evaluate_parameters.py

```diff
@@ -54,28 +66,38 @@ def evaluate_parameter_dict(
             if isinstance(value[0], Substitution):
                 # Value is a list of substitutions, so perform them to make a string
                 evaluated_value = perform_substitutions(context, list(value))
-                evaluated_value = yaml.safe_load(evaluated_value)
+                yaml_evaluated_value = yaml.safe_load(evaluated_value)
+                if type(yaml_evaluated_value) in (bool, int, float, str, bytes):
+                    evaluated_value = yaml_evaluated_value
+                elif isinstance(yaml_evaluated_value, Sequence):
+                    # str and bytes were already handled in the previous case
+                    # If it is a list with dissimilar types, don't evaluate the value as yaml.
+                    if not check_sequence_type_is_allowed(yaml_evaluated_value):
+                        raise TypeError(
+                            'Expected a non-empty sequence, with items of uniform type. '
+                            'Allowed sequence item types are bool, int, float, str.'
+                        )
+                    evaluated_value = tuple(yaml_evaluated_value)
+                else:
+                    raise TypeError(
+                        'Allowed value types are bytes, bool, int, float, str, Sequence[bool]'
+                        ', Sequence[int], Sequence[float], Sequence[str]. Got {}.'.format(
+                            type(yaml_evaluated_value)
+                        )
+                    )
             elif isinstance(value[0], Sequence):
                 # Value is an array of a list of substitutions
                 output_subvalue: List[str] = []
                 for subvalue in value:
                     value = perform_substitutions(context, list(subvalue))
                     output_subvalue.append(value)
-                    evaluated_value = tuple(output_subvalue)
+                evaluated_value = tuple(output_subvalue)
                 # All values in a list must have the same type.
                 # If they don't then assume it is a list of strings
-                yaml_evaluated_value: Union[List[str], List[int], List[float], List[bool]] = []
-                last_subtype = None
-                dissimilar_types = False
-                for subvalue in evaluated_value:
-                    yaml_subvalue = yaml.safe_load(subvalue)
-                    subtype = type(yaml_subvalue)
-                    if last_subtype is not None and last_subtype != subtype:
-                        dissimilar_types = True
-                        break
-                    yaml_evaluated_value.append(yaml_subvalue)
-                    last_subtype = subtype
-                if not dissimilar_types:
+                yaml_evaluated_value: Union[List[str], List[int], List[float], List[bool]] = [
+                    yaml.safe_load(item) for item in evaluated_value
+                ]
+                if check_sequence_type_is_allowed(yaml_evaluated_value):
```

Comment:
```suggestion
                if not check_sequence_type_is_allowed(yaml_evaluated_value):
                    raise TypeError(
                        'Expected a non-empty sequence, with items of uniform type. '
                        'Allowed sequence item types are bool, int, float, str.'
                    )
                evaluated_value = tuple(yaml_evaluated_value)
```

That change will also raise an error in:
https://github.com/ros2/launch_ros/blob/4efbfe5e0933a3747954f78b1c765348711375e9/test_launch_ros/test/test_launch_ros/test_normalize_parameters.py#L214-L239
https://github.com/ros2/launch_ros/blob/4efbfe5e0933a3747954f78b1c765348711375e9/test_launch_ros/test/test_launch_ros/test_normalize_parameters.py#L283-L293

I can add complicated logic for allowing interpreting some dissimilar arrays as arrays of strings, and some others not.
But, what's that useful for?

I think that in all the cases in `test_dictionary_with_dissimilar_array`, the user could write easily the parameters in other wat for getting an array of strings (if they want that). I wouldn't be so flexible as we are being now.

If we want that change, I could later simplify `normalize_parameters` logic in a follow-up PR.

Reply:
That sounds fine to me. It's better to be strict and relax it later if someone makes a case for it, than to make it stricter later.

Reply:
To be clear, +1 to ratcheting up the exception more if all the cases don't throw yet, whether or not that's just this change or something additional.